content/4_Gold/PURS.mdx

```diff
@@ -160,241 +160,300 @@ public class segtree {
-
-		public SegmentTree(int n) {
-			this.n = n;
-			tree = new int[n*2];
-		}
+	public static void main(String[] args) {
+		Kattio io = new Kattio();
+		int arrLen = io.nextInt();
+		int queryNum = io.nextInt();
 
-		public int min(int a, int b) {
-			a += n;
-			b += n;
-			int min = Integer.MAX_VALUE;
-			while (a <= b) {
-				if (a%2 == 1) min = Math.min(min, tree[a++]);
-				if (b%2 == 0) min = Math.min(min, tree[b--]);
-				a /= 2;
-				b /= 2;
-			}
-			return min;
+		MinSegmentTree segtree = new MinSegmentTree(arrLen);
+		for (int i = 0; i < arrLen; i++) {
+			segtree.set(i, io.nextInt());
 		}
 
-		public void set(int index, int val) {
-			index += n;
-			tree[index] = val;
-			for (index /= 2; index >= 1; index /= 2) {
-				tree[index] = Math.min(tree[2*index], tree[2*index+1]);
+		for (int i = 0; i < queryNum; i++) {
+			int type = io.nextInt();
+			if (type == 1) {
+				int ind = io.nextInt() - 1;
+				int val = io.nextInt();
+				segtree.set(ind, val);
+			} else {
+				int start = io.nextInt() - 1;
+				int end = io.nextInt() - 1;
+				io.println(segtree.rangeMin(start, end + 1));
 			}
 		}
+		io.close();
 	}
 
-	CodeSnip{Kattio}
+	//CodeSnip{Kattio}
 }
 ```
 
 </JavaSection>
-
 </LanguageSection>
 
 ### Solution - Dynamic Range Sum Queries
 
 <LanguageSection>
-
 <CPPSection>
 
-Compared to the previous problem, all we need to change are `T`, `ID`, and
-`comb`.
+Compared to the previous problem, all we need to change are `DEFAULT` and `comb`.
 
 ```cpp
-#include <bits/stdc++.h>
-using namespace std;
+#include <iostream>
+#include <cassert>
+#include <vector>
+
+using std::cout;
+using std::endl;
+using std::vector;
+
+template<class T>
+class SumSegmentTree {
+	private:
+		T comb(T a, T b) { return a + b; }
+		const T DEFAULT = 0;
+
+		vector<T> segtree;
+		int len;
+	public:
+		SumSegmentTree(int len) : len(len), segtree(len * 2, DEFAULT) { }
+
+		void set(int ind, T val) {
+			assert(0 <= ind && ind < len);
+			for (segtree[ind += len] = val; ind > 1; ind >>= 1) {
+				segtree[ind >> 1] = comb(segtree[ind], segtree[ind ^ 1]);
+			}
+		}
 
-template<class T> struct Seg { // comb(ID,b) = b
-	const T ID = 0; T comb(T a, T b) { return a+b; }
-	int n; vector<T> seg;
-	void init(int _n) { n = _n; seg.assign(2*n,ID); }
-	void pull(int p) { seg[p] = comb(seg[2*p],seg[2*p+1]); }
-	void upd(int p, T val) { // set val at position p
-		seg[p += n] = val; for (p /= 2; p; p /= 2) pull(p); }
-	T query(int l, int r) {	// sum on interval [l, r]
-		T ra = ID, rb = ID;
-		for (l += n, r += n+1; l < r; l /= 2, r /= 2) {
-			if (l&1) ra = comb(ra,seg[l++]);
-			if (r&1) rb = comb(seg[--r],rb);
+		T range_min(int start, int end) {
+			assert(0 <= start && start < len && 0 < end && end <= len);
+			T sum = DEFAULT;
+			for (start += len, end += len; start < end; start >>= 1, end >>= 1) {
+				if ((start & 1) != 0) {
+					sum = comb(sum, segtree[start++]);
+				}
+				if ((end & 1) != 0) {
+					sum = comb(sum, segtree[--end]);
+				}
```

Comment:
r u sure it's `!= 0` not `== 0`? I mean i don't know much about segT... The original cpp and java code were already different though: 
![image](https://user-images.githubusercontent.com/44814054/184588612-b9f5c53f-ffa3-4ed7-901d-424193ac9142.png)
![image](https://user-images.githubusercontent.com/44814054/184588661-d83fd485-725d-4728-a709-98519351e83e.png)

Reply:
trust me, this works (you can test it yourself)

Reply:
then alright : )